src/ucs/debug/instrument.c

```diff
@@ -110,22 +136,59 @@ void ucs_instrument_init()
 disable_close_file:
     close(ucs_instr_ctx.fd);
 disable:
-    ucs_instr_ctx.enable = 0;
+    ucs_instr_ctx.fd = -1;
+    ucs_instr_ctx.enabled = 0;
     ucs_trace("instrumentation is disabled");
 }
 
 void ucs_instrument_cleanup()
 {
-    if (!ucs_instr_ctx.enable) {
+    if (ucs_instr_ctx.fd == -1) {
         return;
     }
 
     ucs_instrument_write();
     close(ucs_instr_ctx.fd);
-    free(ucs_instr_ctx.start);
+    ucs_free(ucs_instr_ctx.start);
+    ucs_free(ucs_instr_ctx.registered);
+}
+
+uint32_t __ucs_instrument_register(ucs_instrumentation_types_t type,
+                                   char const *location_name)
+{
+    ucs_instrument_location_t *location_entry;
+    ucs_instrument_context_t *ctx = &ucs_instr_ctx;
+
+    if ((ctx->enabled & UCS_BIT(type)) == 0) {
+        return 0;
+    }
+
+    ctx->registered_len++;
+    ctx->registered = ucs_realloc(ctx->registered,
+                                  ((ctx->registered_len / 8) + 1),
+                                  "instrument_registry");
+    if (ctx->registered == NULL) {
+        ctx->registered_len = 0;
+        return 0;
+    }
+
+    location_entry = ucs_calloc(1, sizeof(*location_entry),
+            "instrument_location");
+    if (location_entry == NULL) {
+        return 0;
+    }
+
+    /* Fill in a location entry */
+    strncpy(location_entry->name, location_name,
+            sizeof(location_entry->name) - 1);
+    location_entry->location = ctx->registered_len;
+    ucs_list_add_tail(&ctx->locations_head, &location_entry->list);
+
+    ctx->registered[ctx->registered_len / 8] |= UCS_BIT(ctx->registered_len % 8);
+    return ctx->registered_len;
 }
 
-void __ucs_instrument_record(uint64_t location, uint64_t lparam, uint32_t wparam)
+void __ucs_instrument_record(uint32_t location, uint64_t lparam, uint32_t wparam)
```

Comment:
please remove leading __ (even though the original function had __, we decided later on to remove all those gradually)